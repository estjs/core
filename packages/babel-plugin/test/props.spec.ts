import { getTransform } from './transformUtil';
const transformCode = getTransform('props');
describe('props', () => {
  it('should replace function parameter with "props"', () => {
    const input = `
      function testFunction({ prop1, prop2 }) {
        return <div prop1={prop1} prop2={prop2} />;
      }
    `;

    const code = transformCode(input);
    expect(code).toMatchSnapshot();
  });

  it('should work with rest pattern props', () => {
    const input = `
      function testFunction({ prop1, prop2, ...restProps }) {
        return <div prop1={prop1} prop2={prop2} {...restProps} />;
      }
    `;

    const code = transformCode(input);
    expect(code).toMatchSnapshot();
  });

  it('should work with rest value props', () => {
    const input = `
      function testFunction({ prop1, prop2, ...restProps }) {
        return <div prop1={prop1} prop2={prop2} rest={{...restProps}} />;
      }
    `;

    const code = transformCode(input);
    expect(code).toMatchSnapshot();
  });

  it('should work with deep props', () => {
    const input = `
      function testFunction({ prop1: { prop2: { prop3: {prop4 ,prop5}} } }) {
        return <div prop4={prop4} prop5={prop5}/>;
      }
    `;

    const code = transformCode(input);

    expect(code).toMatchSnapshot();
  });

  it("should replace function parameter with 'props' when it's an object", () => {
    const input = `
      function testFunction({ prop1, prop2 },otherProps) {
        return <div prop1={prop1} prop2={prop2} >{otherProps}</div>;
      }
    `;

    const code = transformCode(input);
    expect(code).toMatchSnapshot();
  });
  it('should not work with array props', () => {
    const input = `
      function testFunction([prop1, prop2],otherProps) {
        return <div prop1={prop1} prop2={prop2} >{otherProps}</div>;
      }
    `;

    const code = transformCode(input);
    expect(code).toMatchSnapshot();
  });
  it('should not work with not pattern props', () => {
    const input = `
      function testFunction(prop1, prop2) {
        return <div prop1={prop1} prop2={prop2} />;
      }
    `;

    const code = transformCode(input);
    expect(code).toMatchSnapshot();
  });
  it('should work just rest props', () => {
    const input = `
      function testFunction({ ...restProps}) {
        return <div prop1={restProps.$prop1} prop2={restProps.prop2} rest={{
          ...restProps
        }} />;
      }
    `;

    const code = transformCode(input);
    expect(code).toMatchSnapshot();
  });
});
