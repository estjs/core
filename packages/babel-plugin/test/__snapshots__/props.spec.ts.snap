// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`props > should not work with array props 1`] = `
"function testFunction([prop1, prop2], otherProps) {
  return <div prop1={prop1} prop2={prop2}>{otherProps}</div>;
}"
`;

exports[`props > should not work with not pattern props 1`] = `
"function testFunction(prop1, prop2) {
  return <div prop1={prop1} prop2={prop2} />;
}"
`;

exports[`props > should replace function parameter with "props" 1`] = `
"function testFunction(__props) {
  return <div prop1={__props.prop1} prop2={__props.prop2} />;
}"
`;

exports[`props > should replace function parameter with 'props' when it's an object 1`] = `
"function testFunction(__props, otherProps) {
  return <div prop1={__props.prop1} prop2={__props.prop2}>{otherProps}</div>;
}"
`;

exports[`props > should work just rest props 1`] = `
"function testFunction(restProps) {
  return <div prop1={restProps.$prop1} prop2={restProps.prop2} rest={{
    ...restProps
  }} />;
}"
`;

exports[`props > should work with deep props 1`] = `
"function testFunction(__props) {
  return <div prop4={__props.prop1.prop2.prop3.prop4} prop5={__props.prop1.prop2.prop3.prop5} />;
}"
`;

exports[`props > should work with rest pattern props 1`] = `
"import { reactive as _reactive$ } from "essor";
function testFunction(__props) {
  const restProps = _reactive$(__props, ["prop1", "prop2"]);
  return <div prop1={__props.prop1} prop2={__props.prop2} {...restProps} />;
}"
`;

exports[`props > should work with rest value props 1`] = `
"import { reactive as _reactive$ } from "essor";
function testFunction(__props) {
  const restProps = _reactive$(__props, ["prop1", "prop2"]);
  return <div prop1={__props.prop1} prop2={__props.prop2} rest={{
    ...restProps
  }} />;
}"
`;
